public/content/lessons/2024/gpt/index.mdx

```diff
@@ -1,9 +1,231 @@
+
+### Embedding Matrix
+
+The model has a predefined vocabulary, some list of all possible words, say 50,000 of them, and the first matrix of the transformer, known as the **embedding matrix**, will have one column for each of these words. These columns determine what vector each word turns into in that first step. We label it as $W_E$, and like all the matrices we see, its values begin random, but they're going to be learned based on data.
+
+Turning words into vectors was common practice in machine learning long before transformers, and it sets the foundation for everything that follows, so let’s take a moment to get familiar with it. This is often called _embedding the word_, which invites thinking of these vectors geometrically as points or directions in some space. Visualizing a list of three numbers as coordinates for a point in 3D space is no problem, but word embeddings tend to be very high dimensional. For GPT-3, they have 12,288 dimensions, and as seen, it matters to work in a space with lots of distinct directions.
+
+<Figure image="embedding.png" width="100%"/>
+
+In the same way that you can take a 2D slice through 3D space and project points onto that slice, for the sake of visualizing word embeddings, we will do something analogous by choosing a 3D _slice_ through the very high-dimensional space, projecting word vectors onto that, and displaying the result.
+
+### Direction
+
+The big idea we need to understand here is that as a model tweaks and tunes its weights to decide how exactly words get embedded as vectors during training, it tends to settle on a set of embeddings where directions in this space have semantic meaning. Below, a simple word-to-vector model is running, and when I run a search for all words whose embeddings are closest to that of _tower_, they all generally have the same vibe.
+
+<Figure image="tower.png" width="100%"/>
+
+Another classic example of this is when the difference between the vectors for _woman_ and _man_ is taken, which can be visualized as a vector in this space connecting the tip of one to the tip of the other. This difference is quite similar to the difference between _king_ and _queen_. So, if the word for a female monarch was unknown, it could be found by taking _king_, adding the direction of _woman_ minus _man_, and searching for the closest word embedding.
+
+<Figure image="queen.png" width="100%"/>
+
+At least, kind of. Despite this being the classic example, for this model the true embedding of _queen_ is a little farther off than the difference would suggest, presumably because the way _queen_ is used in training data is not merely a feminine version of a king. Family relations illustrate the idea better:
```

Comment:
```suggestion
At least, kind of. Despite this being the classic example, for the model I was playing with when making this video, the true embedding of _queen_ is a little farther off than the difference would suggest, presumably because the way _queen_ is used in training data is not merely a feminine version of a king. Family relations illustrate the idea better:
```